clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2223,6 +2288,10 @@ namespace {
             BoundsContextTy InitialObservedBounds = BlockState.ObservedBounds;
             BlockState.G.clear();
 
+            // If any bounds are killed by statement S, remove their bounds
+            // from the ObservedBounds.
+            ResetKilledBounds(KilledBounds, S, BlockState.ObservedBounds);
```

Comment:
I think this should be done before saving `InitialObservedBounds = BlockState.ObservedBounds`. `InitialBounds` should contain the contents of `ObservedBounds` before calling `Check(S, CSS, BlockState)` in order for the observed bounds to be reset after the `Check` call.

For example:
```
void pointer_widening() {
  nt_array_ptr<char> p : count(0) = "a";

  if (p[0]) {
    // This assignment kills the widened bounds of p.
    // InitialObservedBounds should contain p => bounds(p, p + 0),
    // not p => bounds(p, p + 0 + 1).
    // After checking this assignment, BlockState.ObservedBounds should contain
    // p => bounds(p, p + 0).
    p = "";
    if (p[1]) // observed bounds of p are (p, p + 0), so this is an out-of-bounds memory access
  }
}
``